core/shared/src/main/scala/zio/internal/metrics/ConcurrentMetricRegistry.scala

```diff
@@ -42,6 +44,43 @@ private[zio] class ConcurrentMetricRegistry {
     } else hook0.asInstanceOf[Result]
   }
 
+  def addListener(keyType: MetricKeyType)(listener: MetricListener): Unit =
+    listeners.compute(
+      keyType,
+      { case (_, listeners) =>
+        listeners match {
+          case null      => Set(listener)
+          case listeners => listeners + listener
+        }
+      }
+    )
+
+  def removeListener(listener: MetricListener): Unit =
+    listeners.keySet().forEach { metricKeyType =>
+      listeners.computeIfPresent(metricKeyType, { case (a, b) => b - listener })
+    }
+
+  def update[T](key: MetricKey[MetricKeyType.WithIn[T]], value: T)(implicit trace: Trace): Unit = {
+    val listenersForKey = listeners.get(key.keyType)
+    val iterator        = listenersForKey.iterator
```

Comment:
Here you are always allocating an `Iterator`, together with a `Function0`. At the very least, you should fast-path this by checking to see if `listeners.nonEmpty` before executing all of this (heavyweight) logic.